### DISCUSSION TITLE
What is the correct replacement for combineLatest inside a pipe chain?

### DISCUSSION THREAD

**Discussion by mikesnare** [ReactiveX/rxjs#6230] [Q&A] [answered]
If ALL of the combineLatest operator overloads have been deprecated, what is the expected replacement when you're in the middle of a pipe chain?  Imagine you currently have this:

```ts
a$.pipe(
  filter(...),
  switchMap(...),
  map(...),
  etc(...),
  combineLatest([b$, c$])
)
```

First... _why is that wrong?_  What's so bad about it that it had to be deprecated?  But moving on...

What's the correct replacement?  Is it expected that we restructure that to this?
```ts
combineLatest([
  a$.pipe(
    filter(...),
    switchMap(...),
    map(...),
    etc(...)
  ),
  b$,
  c$
])
```

or this?
```ts
a$.pipe(
  filter(...),
  switchMap(...),
  map(...),
  etc(...),
  switchMap(val => combineLatest([of(val), b$, c$]))
)
```
I'm not even sure that last one will work?  If `b$` and `c$` don't replay their last emissions I don't think it's a valid replacement.

**Comment by cartant** [comment, marked as answer]
The replacement should be named in the deprecation message:

https://github.com/ReactiveX/rxjs/blob/df51b04d7ec68a72b3a4b0d69c3bb29264c72611/src/internal/operators/combineLatest.ts#L9

FWIW, the deprecation messages are being worked on ATM. When done, each will contain a link to a document that will explain in detail the reason for the deprecation and the changes that developers need to make to avoid relying upon what's deprecated.

**Reply by cartant** [reply]
Actually, it's in v7 that there is a sensible deprecation message. The v6 message isn't super helpful:

https://github.com/ReactiveX/rxjs/blob/01fcffc6276f13761d5d455055edad3c33b4684e/src/internal/operators/combineLatest.ts#L11

The reason for the deprecation was that the operator had the same name as the static creator - which wasn't a problem in v5, as the operators were on the `Observable` prototype. In v6, I guess the 'correct' replacement is/was to use the static creator:

```ts
a$.pipe(
  filter(...),
  switchMap(...),
  map(...),
  etc(...),
  s$ => combineLatest([s$, b$, c$])
)
```

**Reply by mikesnare** [reply]
Ok.  There's the issue.  We're on 6, not 7, so our messages only referenced the static version, and combineLatestWith doesn't exist until release 7.  Thanks for the reply.
